### PR TITLE
Add security_hal features

### PR DESCRIPTION
**Pull request by SeonghoByeon**
#### commit id: 25305a5
- Provide a generic format of security_hal (security_hal.h)

#### commit id: 4dfd39d 
apps/examples/security_test: Refactor security_test example features
- Add security_hal testcases
- Refactor security test in apps/examples/security_test for see_api testcases

#### commit id: ddb0f5c
- os/arch/s5j: Add a security hal wrapper for Artik isp

### CODE REVIEW

**Comment by SeonghoByeon** [comment]
Have successfully tested security hal test case over Artik SSS (using wrapper which translates hal to isp APIs).

**Comment by seinfra** [comment]
Target : [ddb0f5c2ff534ed8ec8291a705f7ab0f5e9a39ad] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ddb0f5c2ff534ed8ec8291a705f7ab0f5e9a39ad] - Code Rule Check Result:
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:33: ERROR: [IDT_M_TAB] please, no space before tabs
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:99: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:99: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:100: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:100: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:102: ERROR: [SPC_M_SEP] space required after that close brace '}'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:103: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:103: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:113: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:113: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:113: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:232: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:333: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:370: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:393: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:424: ERROR: [SPC_M_OPR] spaces required around that '==' (ctx:WxV)
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:522: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:576: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:576: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:586: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:586: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:586: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:586: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:605: ERROR: [SPC_M_OPR] space prohibited before that ':' (ctx:WxE)
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:623: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:623: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:623: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:644: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:708: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:801: ERROR: [SPC_M_OPR] spaces required around that '!=' (ctx:WxV)
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:821: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:925: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:980: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1024: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1041: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1069: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1069: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1070: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1070: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1072: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1072: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1077: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1077: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1084: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1084: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1085: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1085: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1092: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1092: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1093: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1093: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1109: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1109: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1110: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1110: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1112: ERROR: [SPC_M_OPR] spaces required around that '!=' (ctx:VxW)
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1112: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1122: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1123: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1146: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1164: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1214: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1264: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1303: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1463: ERROR: [IDT_M_SCH] switch and case should be at the same indent
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1511: ERROR: [IDT_M_SCH] switch and case should be at the same indent

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 1]
```diff
@@ -0,0 +1,1627 @@
...
+{
+	if ((end - *p) < 1) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if ((**p & 0x80) == 0) {
+		*len = *(*p)++;
+	} else {
+		switch (**p & 0x7F) {
+			case 1:
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Think you can clean this by making 1 simple loop.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 2]
```diff
@@ -0,0 +1,1627 @@
...
+	}
+
+	/* Signature encoding */
+	hal_mpi r, s;
+	hal_mpi_init( &r );
+	hal_mpi_init( &s );
+
+	ret = hal_mpi_read_binary( &r, ecc_sign.r, ecc_sign.r_byte_len );
+	if (ret != 0) {
+		goto cleanup;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: why you didn't use your MACRO?

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 3]
```diff
@@ -0,0 +1,1627 @@
...
+	ret = isp_ecdsa_sign_md_securekey(&ecc_sign, hash->data, hash->data_len, key_idx);
+	if (ret != 0) {
+		isp_clear(0);
+		printf("ISP failed (%zu)\n", ret);
+		return HAL_FAIL;
+	}
+
+	/* Signature encoding */
+	hal_mpi r, s;
+	hal_mpi_init( &r );
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: bracket.

File: apps/examples/security_test/see_api/see_test_main.c [hunk 4]
```diff
@@ -34,9 +34,9 @@
 /*
  * Definition for handling pthread
  */
-#define SECURITY_TEST_PRIORITY     100
-#define SECURITY_TEST_STACK_SIZE   51200
-#define SECURITY_TEST_SCHED_POLICY SCHED_RR
+#define SEE_TEST_PRIORITY     100
+#define SEE_TEST_STACK_SIZE   51200
+#define SEE_TEST_SCHED_POLICY SCHED_RR
```
Comment (Taejun-Kwon) [inline apps/examples/security_test/see_api/see_test_main.c]: sched.h should be included.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 5]
```diff
@@ -0,0 +1,1627 @@
+/****************************************************************************
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: And it using constant many values but almost no description.
Then it is very hard to understand when other ppl see yours.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 6]
```diff
@@ -0,0 +1,1632 @@
...
+
+static int hal_asn1_write_tag(unsigned char **p, unsigned char *start, unsigned char tag)
+{
+	if (*p - start < 1) {
+		return -HAL_INVALID_ARGS;
+	}
+
+	*--(*p) = tag;
+
+	return 1;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: some functions return HAL_SUCCESS(c.f. hal_mpi_write_binary), here rutnrs 1;
Why don't you make return types first?
Same opinion for error case.

**Comment by seinfra** [comment]
Target : [ac958912482689b732c5cc34a90ddd72797623d9] - Code Rule Check (C++) OK.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 7]
```diff
@@ -0,0 +1,1632 @@
...
+			return HAL_NOT_ENOUGH_MEMORY;
+		}
+		*--(*p) = 0x00;
+		len += 1;
+	}
+
+	HAL_ASN1_CHK_ADD(len, hal_asn1_write_len(p, start, len));
+	HAL_ASN1_CHK_ADD(len, hal_asn1_write_tag(p, start, 0x02));
+
+	ret = (int)len;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Is it intended type?

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 8]
```diff
@@ -0,0 +1,1632 @@
...
+	len = hal_mpi_size(X);
+
+	if (*p < start || (uint32_t)(*p - start) < len) {
+		return -HAL_INVALID_ARGS;
+	}
+
+	(*p) -= len;
+	HAL_MPI_CHK(hal_mpi_write_binary(X, *p, len));
+
+	if (X->s == 1 && **p & 0x80) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: please add bracket set.

**Comment by seinfra** [comment]
Target : [ac958912482689b732c5cc34a90ddd72797623d9] - Code Rule Check Result:
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:106: ERROR: [SPC_M_SEP] space required after that close brace '}'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:807: ERROR: [SPC_M_OPR] spaces required around that '!=' (ctx:WxV)
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:807: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1119: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
os/arch/arm/src/s5j/sss/security_hal_wrapper.c:1129: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 9]
```diff
@@ -0,0 +1,1632 @@
...
+	} else if (key_idx == FACTORYKEY_ARTIK_DEVICE) {
+		ISP_CHECKBUSY();
+		if (!key->data) {
+			return HAL_INVALID_ARGS;
+		}
+		key->data_len = HAL_MAX_BUF_SIZE;
+		ret = isp_get_factorykey_data(key->data, &key->data_len, key_idx);
+		if (ret != 0) {
+			isp_clear(0);
+			printf("ISP failed (%zu)\n", ret);
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: please use debug

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 10]
```diff
@@ -0,0 +1,1632 @@
...
+ * Key Manager
+ */
+int hal_set_key(hal_key_type mode, uint32_t key_idx, hal_data *key, hal_data *prikey)
+{
+	HWRAP_ENTER;
+	if (prikey != NULL) {
+		return HAL_NOT_SUPPORTED;
+	}
+
+	unsigned int key_type;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: How about rename 'hal_get_key_type' to hal_get_ss_type_from_key_type
and then use it here. and second param is not neccessary, you can change return type.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 11]
```diff
@@ -0,0 +1,1632 @@
...
+	}
+
+	return HAL_SUCCESS;
+}
+
+int hal_get_key(hal_key_type mode, uint32_t key_idx, hal_data *key)
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	if (mode <= HAL_KEY_ECC_SEC_P512R1 && mode >= HAL_KEY_ECC_BRAINPOOL_P256R1) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: bracket set.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 12]
```diff
@@ -0,0 +1,1632 @@
...
+			return HAL_NOT_SUPPORTED;
+		}
+
+		ecc_key.curve = object_id;
+
+		ISP_CHECKBUSY();
+		ret = isp_ecdsa_get_publickey_securekey(&ecc_key, key_idx, object_id);
+		if (ret != 0) {
+			isp_clear(0);
+			printf("ISP failed (%zu)\n", ret);
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: please check all printf here.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 13]
```diff
@@ -0,0 +1,1632 @@
...
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	unsigned int inbuf[HAL_MAX_RANDOM_SIZE];
+
+	if (len > HAL_MAX_RANDOM_SIZE) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if (len & 0x3) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: I mean code like this.
Here it calculates with constant value, it is very hard to understand.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 14]
```diff
@@ -0,0 +1,1632 @@
...
+
+	HAL_COPY_DATA(hash, output, hash_len);
+
+	return HAL_SUCCESS;
+}
+
+int hal_get_hmac(hal_hmac_type mode, hal_data *input, uint32_t key_idx, hal_data *hmac)
+{
+	HWRAP_ENTER;
+	/* isp_hmac_securekey returns ERROR_SSTORAGE_SFS_FREAD */
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: please remove unusing code or add TODO

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 15]
```diff
@@ -0,0 +1,1632 @@
...
+
+	struct sECC_SIGN ecc_sign;
+	memset(&ecc_sign, 0, sizeof(struct sECC_SIGN));
+
+	ecc_sign.s = s_buf;
+	ecc_sign.s_byte_len = HAL_MAX_ECP_KEY_SIZE_ALT;
+	ecc_sign.r = r_buf;
+	ecc_sign.r_byte_len = HAL_MAX_ECP_KEY_SIZE_ALT;
+	
+	switch (mode->hash_t) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: I think you can make function to handle this switch state to reduce code line.

**Comment by seinfra** [comment]
Target : [f8891b77cbbb1e113057c3446e2c3c2e07c587fe] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f8891b77cbbb1e113057c3446e2c3c2e07c587fe] - Code Rule Check OK.

File: apps/examples/security_test/hal/hal_ss_test.c [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+		ST_EXPECT_EQ2(HAL_SUCCESS, HAL_NOT_SUPPORTED, hal_write_storage(1, &g_input));
+	}
+
+	ST_END_TEST;
+}
+
+TEST_TEARDOWN(delete_storage)
+{
+	ST_START_TEST;
+
```
Comment (thapav) [inline apps/examples/security_test/hal/hal_ss_test.c]: Is this TEST_TEARDOWN call required ?

File: apps/examples/security_test/hal/hal_keymgr_test.c [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+/*
+ * Desc: Generate key
+ * Refered https://developer.artik.io/documentation/security-api/see-authentication-test_8c-example.html
+ */
+TEST_SETUP(generate_key)
+{
+	ST_START_TEST;
+
```
Comment (thapav) [inline apps/examples/security_test/hal/hal_keymgr_test.c]: same is this SETUP required here ?

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 18]
```diff
@@ -0,0 +1,1632 @@
...
+{
+	int ret;
+
+	HAL_MPI_CHK(hal_mpi_grow(X, 1));
+	memset(X->p, 0, X->n * ciL);
+
+	X->p[0] = (z < 0) ? -z : z;
+	X->s    = (z < 0) ? -1 : 1;
+
+cleanup:
```
Comment (thapav) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Where is this cleanup used ?

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 19]
```diff
@@ -0,0 +1,1632 @@
...
+		*--(*p) = 0x00;
+		len += 1;
+	}
+
+	HAL_ASN1_CHK_ADD(len, hal_asn1_write_len(p, start, len));
+	HAL_ASN1_CHK_ADD(len, hal_asn1_write_tag(p, start, 0x02));
+
+	ret = (int)len;
+
+cleanup:
```
Comment (thapav) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: same as above

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 20]
```diff
@@ -0,0 +1,1632 @@
...
+		HAL_MPI_CHK(hal_mpi_grow(X, limbs));
+	}
+
+	HAL_MPI_CHK(hal_mpi_lset(X, 0));
+
+	for (i = buflen, j = 0; i > 0; i--, j++) {
+		X->p[j / ciL] |= ((uint64_t)buf[i - 1]) << ((j % ciL) << 3);
+	}
+
+cleanup:
```
Comment (thapav) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: same as above

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 21]
```diff
@@ -0,0 +1,1632 @@
...
+		hash_len = 128;
+		break;
+	default:
+		return HAL_NOT_SUPPORTED;
+	}
+
+	unsigned char output[hash_len];
+	memset(&h_param, 0, sizeof(struct sHASH_MSG));
+
+	h_param.addr_low = (unsigned int) input->data;
```
Comment (thapav) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: remove space after typecast*

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 22]
```diff
@@ -0,0 +1,1632 @@
...
+	hal_mpi_free(&s);
+
+	return HAL_FAIL;
+}
+
+int hal_ecdsa_verify_md(hal_ecdsa_mode mode, hal_data *hash, hal_data *sign, uint32_t key_idx)
+{
+	HWRAP_ENTER;
+	uint32_t ret;
+	unsigned char *p = (unsigned char *) sign->data;
```
Comment (thapav) [inline os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: remove space after typecast

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 15 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Thanks for the comment. Unfortunately, specific conditions within the switch statement differ slightly for each case. In addition, curve types for specific modules supported by the SE may vary.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 14 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Applied. Thanks.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 13 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: This format is provided by DS isp code, which we do not need to take care. Thanks for the comment.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 12 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Applied. Thanks for the comment.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 11 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Applied. Thanks for the comment.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 1 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: This format is standard ASN1 encoding param provided by mbedtls.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 23]
```diff
@@ -0,0 +1,1627 @@
...
+static int hal_asn1_get_len(unsigned char **p, const unsigned char *end, uint32_t *len)
+{
+	if ((end - *p) < 1) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if ((**p & 0x80) == 0) {
+		*len = *(*p)++;
+	} else {
+		switch (**p & 0x7F) {
```
[the remark this answers is not in the document]
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: This format is standard ASN1 encoding param provided by mbedtls.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c [hunk 24]
```diff
@@ -0,0 +1,1627 @@
...
+#define HAL_COPY_PRIV_DATA(in, out, len)                \
+	do {                                                \
+		if (in->priv == NULL) {                         \
+			return HAL_INVALID_ARGS;                    \
+		}                                               \
+		memcpy(in->priv, out, len);                     \
+		in->priv_len = len;                             \
+	} while (0)
+
+/**
```
[the remark this answers is not in the document]
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Applied. Thanks.

File: apps/examples/security_test/hal/hal_test_utils.c [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <tinyara/config.h>
+#include <tinyara/security_hal.h>
+
+#include <stdlib.h>
+#include "hal_test_utils.h"
+
+void hal_test_free_buffer(hal_data *data)
+{
+	if (data->data) {
```
[the remark this answers is not in the document]
Comment (SeonghoByeon) [reply apps/examples/security_test/hal/hal_test_utils.c]: That's good for safety. Thanks.

File: apps/examples/security_test/hal/hal_test_main.c [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+	if (res != 0) {
+		printf("Failed to pthread_attr_setstacksize\n");
+		return -1;
+	}
+
+	res = pthread_create(&tid, &attr, (pthread_startroutine_t)hal_keymgr_test, NULL);
+	if (res < 0) {
+		printf("hal key manager test fail\n");
+	}
+	res = pthread_join(tid, NULL);
```
[the remark this answers is not in the document]
Comment (SeonghoByeon) [reply apps/examples/security_test/hal/hal_test_main.c]: You are right. Thanks for the comment.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 22 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Removed. Thanks.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 21 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Removed. Thanks.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 20 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Removed. Thanks.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 19 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Removed. Thanks.

File: os/arch/arm/src/s5j/sss/security_hal_wrapper.c (hunk 18 above)
Comment (SeonghoByeon) [reply os/arch/arm/src/s5j/sss/security_hal_wrapper.c]: Removed. Thanks.

File: apps/examples/security_test/hal/hal_keymgr_test.c (hunk 17 above)
Comment (SeonghoByeon) [reply apps/examples/security_test/hal/hal_keymgr_test.c]: It should be defined, following the rule of stress tool. Thanks for the comment.

File: apps/examples/security_test/hal/hal_ss_test.c (hunk 16 above)
Comment (SeonghoByeon) [reply apps/examples/security_test/hal/hal_ss_test.c]: It should be defined, following the rule of stress tool. Thanks for the comment.

**Comment by SeonghoByeon** [comment]
Hi @drashti304, thanks for the comments. 
> Please provide function description as standard we follow in TizenRT
The code defined in security_hal_wrapper.c denotes a simple wrapper to use isp functions on ARTIK board. Specific function description is provided by the security_hal.h. Please refer to it.




> I think default case is needed
What do you mean default? Please let me know details.

BRs,
Seongho Byeon.

**Comment by seinfra** [comment]
Target : [5ccfd39ce717f580e421589aac0f3150a769bff5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5ccfd39ce717f580e421589aac0f3150a769bff5] -

**Comment by seinfra** [comment]
Target : [20f843198f89103cf66caa69e0891aaa27746606] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [20f843198f89103cf66caa69e0891aaa27746606] - Code Rule Check OK.

**Comment by drashti304** [comment]
> Hi @drashti304, thanks for the comments.
> 
> > Please provide function description as standard we follow in TizenRT
> 
> The code defined in security_hal_wrapper.c denotes a simple wrapper to use isp functions on ARTIK board. Specific function description is provided by the security_hal.h. Please refer to it.
> 
> > I think default case is needed
> 
> What do you mean default? Please let me know details.
> 
> BRs,
> Seongho Byeon.

i meant in case() statement

> 

default:
		return HAL_INVALID_ARGS

> 
case was not there.

**Pull request merged**